Add websockets stream api docs (#1074)
onflow/docs: docs/networks/node-ops/access-onchain-data/access-nodes/accessing-data/websockets-stream-api/subscribe-topics.md

---
title: Supported topics
sidebar_label: Supported topics
sidebar_position: 2
---

# Supported Topics  

Below is a list of topics that can be subscribed to in order to receive updates about different states of the Flow blockchain. It is possible to subscribe to each topic multiple times with different configurations based on input arguments. The responses for all topics are aligned with [Flow REST API](/http-api) responses.

## `block_digests` topic

Provides a summarized version of block information, including only the block ID, height, and timestamp, each time a new block appears on the blockchain. 

### Example Request

```json
{
  "subscription_id": "some-id-1",
  "action": "subscribe",
  "topic": "block_digests",
  "arguments": {
    "block_status": "sealed",

>     "start_block_height": "10,530,102"

Reviewer: I don't think we support comas in block heights
```suggestion
    "start_block_height": "10530102"
```

  }
}
```  

#### Request Arguments

| Name                 | Type   | Required | Description                                                                                                                          |
| -------------------- | ------ | -------- | ------------------------------------------------------------------------------------------------------------------------------------ |

> | `block_status`       | STRING | YES      | The status of blocks to subscribe to.                                                                                                |

Reviewer: can you include the supported values

> | `start_block_id`     | STRING | NO       | The ID of the block from which the subscription starts. If this argument is set, it is **not** possible to set `start_block_height`. |  |

Reviewer: same updates in the other endpoints
```suggestion
| `start_block_id`     | STRING | NO       | The ID of the block from which the subscription starts. If this argument is set, `start_block_height` MUST be empty. |  |
```

> | `start_block_height` | STRING | NO       | The height of the block from which the subscription starts. If this argument is set, it is **not** possible to set `start_block_id`  |

Reviewer: ```suggestion
| `start_block_height` | STRING | NO       | The height of the block from which the subscription starts. If this argument is set, `start_block_id` MUST be empty. |
```

        "block_status": "BLOCK_SEALED"
    }
}
```

## `events` topic

Provides blockchain events. The response can be configured using additional arguments to filter and retrieve only filtered events instead of all events. 

### Example Request

```json
{
  "subscription_id": "some-id-4",
  "action": "subscribe",
  "topic": "events",
  "arguments": {
    "start_block_height": "10530103",
    "event_types": ["flow.AccountKeyAdded"]
  }
}
```  

#### Request Arguments

| Name                 | Type   | Required | Description                                                                                                                          |
| -------------------- | ------ | -------- | ------------------------------------------------------------------------------------------------------------------------------------ |
| `start_block_id`     | STRING | NO       | The ID of the block from which the subscription starts. If this argument is set, it is **not** possible to set `start_block_height`. |  |
| `start_block_height` | STRING | NO       | The height of the block from which the subscription starts. If this argument is set, it is **not** possible to set `start_block_id`  |

> | `heartbeat_interval` | STRING | NO       | Interval in block heights at which the server should return a heartbeat message to the client.                                       |

Reviewer: ```suggestion
| `heartbeat_interval` | STRING | NO       | Maximum number of blocks between messages after which a response with no events is returned. This helps the client track progress for sparse event filters. |
```

> | `event_types`        | LIST   | NO       | A comma-separated list of events type to include.                                                                                    |

Reviewer: ```suggestion
| `event_types`        | LIST   | NO       | A comma-separated list of event types to include.                                                                                    |
```

| `addresses`          | LIST   | NO       | A comma-separated list of addresses who's events should be included.                                                                 |
| `contracts`          | LIST   | NO       | A comma-separated list of contracts who's events should be included.                                                                 |

If neither `start_block_id` nor `start_block_height` is set, the subscription will start from the latest block based on its status.

> ### Example Responce

Author: Responce => Response. here and in other places

Reviewer 2: Yeah we should correct those spelling errors thanks

```json
{
  "subscription_id": "some-id-4",
  "topic": "events",
  "payload":  {

>         "block_id": "0385...",

Reviewer: does blockID include a `0x` prefix?

}
```

## `account_statuses` topic

Provides accounts statuses updates. The response can be configured using additional arguments to filter and retrieve only filtered account statuses instead of all core account events.

### Example Request

```json
{
  "subscription_id": "some-id-5",
  "action": "subscribe",
  "topic": "account_statuses",
  "arguments": {
    "start_block_height": "10530103",
    "event_types": ["flow.AccountCreated"]
  }
}
```  

#### Request Arguments

| Name                 | Type   | Required | Description                                                                                                                          |
| -------------------- | ------ | -------- | ------------------------------------------------------------------------------------------------------------------------------------ |
| `start_block_id`     | STRING | NO       | The ID of the block from which the subscription starts. If this argument is set, it is **not** possible to set `start_block_height`. |  |
| `start_block_height` | STRING | NO       | The height of the block from which the subscription starts. If this argument is set, it is **not** possible to set `start_block_id`  |

> | `heartbeat_interval` | STRING | NO       | Interval in block heights at which the server should return a heartbeat message to the client.                                       |

Reviewer: ```suggestion
| `heartbeat_interval` | STRING | NO       | Maximum number of blocks between messages after which a response with no events is returned. This helps the client track progress for sparse event filters. |
```

> | `event_types`        | LIST   | NO       | A comma-separated list of events type to include.                                                                                    |

Reviewer: ```suggestion
| `event_types`        | LIST   | NO       | A comma-separated list of event types to include.                                                                                    |
```

   }
}
```

## `send_and_get_transaction_statuses` topic

Sends a transaction and provides updates on its status changes.

### Example Request

```json
{
  "subscription_id": "some-id-7",
  "action": "subscribe",
  "topic": "send_and_get_transaction_statuses",
  "arguments": {
    "script": "access(all) fun main() {}",
    "arguments": [],
    "reference_block_id": "a3f1c4d5...",
    "gas_limit": "1000",
    "payer": "01cf...",
    "proposal_key": {
      "address": "01cf...",
      "key_index": 0,
      "sequence_number": 42
    },
    "authorizers": [
      "01cf...",
      "179b..."
    ],
    "payload_signatures": [
        //...
    ],
    "envelope_signatures": [
      //...
    ]
  }
}
```  

| Name                  | Type   | REQUIRED | Description                                                               |
| --------------------- | ------ | -------- | ------------------------------------------------------------------------- |
| `script`              | STRING | YES      | Base64-encoded content of the Cadence script.                             |
| `arguments`           | LIST   | YES      | A list of arguments, each encoded as Base64.                              |

> | `reference_block_id`  | STRING | YES      | A 32-byte unique identifier for an entity.                                |

Reviewer: ```suggestion
| `reference_block_id`  | STRING | YES      | BlockID for the transaction's reference block |
```
